omnipaxos/src/sequence_paxos/leader.rs

```diff
@@ -205,58 +205,59 @@ where
+            accepted_idx: followers_accepted_idx,
             pid,
             ..
         } = self.leader_state.get_promise_meta(to);
-        let follower_decided_idx = self
+        let followers_decided_idx = self
             .leader_state
             .get_decided_idx(*pid)
             .expect("Received PromiseMetaData but not found in ld");
-        let (delta_snapshot, suffix, sync_idx) =
-            if (promise_n == max_promise_n) && (promise_accepted_idx < max_accepted_idx) {
-                if self.internal_storage.get_compacted_idx() > *promise_accepted_idx
-                    && T::Snapshot::use_snapshots()
-                {
-                    let delta_snapshot = self
-                        .internal_storage
-                        .create_diff_snapshot(follower_decided_idx, my_decided_idx)
-                        .expect("storage error while trying to read diff snapshot");
-                    let suffix = self
-                        .internal_storage
-                        .get_suffix(my_decided_idx)
-                        .expect("storage error while trying to read log suffix");
-                    (Some(delta_snapshot), suffix, follower_decided_idx)
-                } else {
-                    let sfx = self
-                        .internal_storage
-                        .get_suffix(*promise_accepted_idx)
-                        .expect("storage error while trying to read log suffix");
-                    (None, sfx, *promise_accepted_idx)
-                }
-            } else if follower_decided_idx < my_decided_idx && T::Snapshot::use_snapshots() {
-                let delta_snapshot = self
+        let (delta_snapshot, suffix, sync_idx) = if (followers_promise_n == max_promise_n) || (*followers_promise_n == self.leader_state.n_leader) {
+            // Follower's accepted entries are valid, send any new entries after
+            if T::Snapshot::use_snapshots() && *followers_accepted_idx < my_decided_idx {
+                // Note: we snapshot from follower's decided and not follower's accepted because
+                // snapshots currently can't handle merging onto accepted entries.
+                let (delta_snapshot, compacted_idx) = self
                     .internal_storage
-                    .create_diff_snapshot(follower_decided_idx, my_decided_idx)
+                    .create_diff_snapshot(followers_decided_idx)
                     .expect("storage error while trying to read diff snapshot");
                 let suffix = self
                     .internal_storage
                     .get_suffix(my_decided_idx)
                     .expect("storage error while trying to read log suffix");
-                (Some(delta_snapshot), suffix, follower_decided_idx)
+                (delta_snapshot, suffix, compacted_idx)
             } else {
+                let sfx = self
+                    .internal_storage
+                    .get_suffix(*followers_accepted_idx)
```

Comment:
This is incorrect. With the check of just `if (followers_promise_n == max_promise_n) || (*followers_promise_n == self.leader_state.n_leader)` before here, it becomes incorrect to send from the followers accepted index. Consider the following case: the old leader, call it server A, disconnects with the other servers at ballot `n=1` and with 10 log entries. The other servers elect a new leader at `n=2` and decides 10 new entries (i.e., entry 0-9 were from `n=1`, and 10-19 were from `n=2`). But during the partition, A continued in `n=1` and appended 5 **other** entries (it will append because it thinks it's the leader). Now when it reconnects in `n=2`, it will keep those 5 entries that were added during the partition, which is incorrect (because its log entries from 10-14 will be from `n=1` instead of `n=2`).

Instead, we should send the `max_accepted_idx` (which you removed). It is only up to that point in the log that we know that this follower's log is valid.
```suggestion
                    .get_suffix(max_accepted_idx)
```

Reply:
Yeah you are right. If the follower was synced in the previous round we can only trust up to the `max_accepted_idx` of last round. And if the follower was synced by us then we can trust up to their `accepted_idx`.